src/text/raw_text_reader.rs

```diff
@@ -760,4 +727,221 @@ mod reader_tests {
 
         Ok(())
     }
+
+    use std::cell::RefCell;
+    use std::rc::Rc;
+
+    // Simple struct to track metrics around reading.
+    struct ReaderMetrics {
+        bytes_read: usize,
+    }
+
+    // An ion data source for use in testing when and how much data is read during parsing.
+    struct TestIonSource {
+        data: &'static str,
+        current: usize,
+        metrics: Rc<RefCell<ReaderMetrics>>,
+    }
+
+    impl TestIonSource {
+        fn new(data: &'static str) -> (TestIonSource, Rc<RefCell<ReaderMetrics>>) {
+            let metrics = Rc::new(RefCell::new(ReaderMetrics { bytes_read: 0 }));
+
+            (
+                TestIonSource {
+                    data,
+                    metrics: metrics.clone(),
+                    current: 0,
+                },
+                metrics,
+            )
+        }
+    }
+
+    impl crate::data_source::ToIonDataSource for TestIonSource {
+        type DataSource = Self;
+
+        fn to_ion_data_source(self) -> Self::DataSource {
+            self
+        }
+    }
+
+    impl std::io::BufRead for TestIonSource {
```

Comment:
Could `TestIonSource` be replaced by [`io::Cursor`](https://doc.rust-lang.org/std/io/struct.Cursor.html) from the stdlib?

Reply:
You can create a reader that uses a `&mut Cursor` and then get the cursor back when you're done. Personally, I'd also be in favor of having a `Reader::input_source()` or something that lent out a reference to the underlying data source. We don't have to solve that in this PR, though.